docs/ovnkube.1
.TH OVNKUBE "1" "Jan 2018" "ovn-kubernetes" "OVN-KUBERNETES User Commands"
.SH NAME
ovnkube \- ovn utility
.SH SYNOPSIS
.B ovnkube
[\fI\,OPTION\/\fR]... [\fI\,FILE\/\fR]...
.SH DESCRIPTION
.PP
The ovn utility used to confiugre ovn-kubernetes
.SH GLOBAL OPTIONS
.TP
\fB\--cluster-subnets\fR string
A comma separated set of IP subnets and the associated hostsubnet prefix lengths to use for
the cluster (eg, "10.128.0.0/14/23,10.0.0.0/14/23").  Each entry is given in the form
[IP address/prefix-length/hostsubnet-prefix-length] and cannot overlap with other entries.
The hostsubnet-prefix-length is optional and if unspecified defaults to 24. The
hostsubnet-prefix-length defines how many IP addresses are dedicated to each node
and may be different for each entry. (default "10.128.0.0/14/23")
.TP
\fB\--k8s-service-cidr\fR value
A CIDR notation IP range from which k8s assigns service cluster IPs.
This should be the same as the one provided for kube-apiserver's
\fB\--service-cluster-ip-range\fR option.
.TP
\fB\--init-master\fR string
Initialize master that acts as a controller that watches pods/services/policies. Requires the hostname as argument.
.TP
\fB\--init-node\fR string
Initialize node, requires the name that node is registered with in kubernetes cluster.
.TP
\fB\--remove-node\fR string
Remove a node from the OVN cluster, requires the name that node is registered
with in kubernetes cluster.
.TP
\fB\--gateway-mode\fR string
On nodes, creates a gateway for traffic to exit the cluster. Set to one of
"shared" or "local". Shared mode shares the gateway interface with
the OVN logical network. Local mode creates a NAT-ed interface for use
with the OVN logical network.
.TP
\fB\--init-gateways\fR
DEPRECATED; use \fB\--gateway-mode\fR instead.
.TP
\fB\--gateway-interface\fR string
The interface in nodes that will be the gateway interface.  If none
specified, then the node's interface on which the default gateway is
configured will be used as the gateway interface. Only useful with
\fB--init-gateways\fR.
.TP
\fB\--gateway-nexthop\fR string
The external default gateway which is used as a next hop by
OVN gateway. This is many times just the default gateway
of the node in question. If not specified, the default gateway
configured in the node is used. Only useful with \fB--init-gateways\fR.
.TP
\fB\--gateway-local\fR
DEPRECATED; use \fB\--gateway-mode\fR instead.
.TP
\fB\--nodeport\fR
Setup nodeport based entries in OVN gateways for ingress into the k8s cluster.
By default, it is disabled.
.TP
\fB\--gateway-v4-join-subnet\fR string
The v4 join subnet to use for assigning join switch IPv4 addresses\fR.
.TP
\fB\--gateway-v6-join-subnet\fR string
The v6 join subnet to use for assigning join switch IPv6 addresses\fR.
.TP
\fB\--gateway-router-subnet\fR string
The Subnet to be used for the gateway router external port (shared mode only). auto-detected if not given.
Currently valid for Smart-NICs only.\fR.
.TP
\fB\--config-file\fR string
Configuration file path.
.TP
\fB\--mtu\fR value
MTU value used for the overlay networks. (default: 0).
.TP
\fB\--conntrack-zone\fR value
For gateway nodes, the conntrack zone used for conntrack flow rules (default: 0).
.TP
\fB\--loglevel\fR int
Log verbosity and level: 5=debug, 4=info, 3=warn, 2=error, 1=fatal (default: 0).
.TP
\fB\--logfile\fR string
Logfile name (with path) for ovnkube to write to.
.TP
\fB\--pidfile\fR string
(Optional) file name that will contain the pid.
.TP
\fB\--cni-conf-dir\fR string
The CNI config directory in which to write the overlay CNI config file.
.TP
\fB\--cni-plugin\fR string
The name of the CNI plugin.
.TP
\fB\--k8s-kubeconfig\fR string
Absolute path to the kubeconfig file (not required if the --k8s-apiserver, --k8s-cacert, and --k8s-token are given).
.TP
\fB\--k8s-apiserver\fR string
URL to the Kubernetes apiserver (default "https://localhost:8443") (not required if --k8s-kubeconfig is given).
.TP
\fB\--k8s-cacert\fR string
The absolute path to the Kubernetes API CA certificate (not required if --k8s-kubeconfig is given).
.TP
\fB\--k8s-token\fR string
The Kubernetes API authentication token (not required if --k8s-kubeconfig is given).
.TP
\fB\--metrics-bind-address\fR string
The IP address and port for the metrics server to serve on (set to 0.0.0.0 for all IPv4 interfaces).
.TP
\fB\--nb-address\fR string
IP address and port of the OVN northbound API (eg, ssl:1.2.3.4:6641). Leave empty to use a local unix socket.
.TP
\fB\--nb-client-privkey\fR string
Private key that the client should use for talking to the OVN database.  Leave empty to use local unix socket.
.TP
\fB\--nb-client-cert\fR string
Client certificate that the client should use for talking to the OVN database.  Leave empty to use local unix socket.
.TP
\fB\--nb-client-cacert\fR string
CA certificate that the client should use for talking to the OVN database.  Leave empty to use local unix socket.
.TP
\fB\--nb-cert-common-name\fR string
The Common Name of the certificate used for TLS server certificate verification.
.TP
\fB\--sb-address\fR string
IP address and port of the OVN southbound database (eg, ssl:1.2.3.4:6642).  Leave empty to use a local unix socket.
.TP
\fB\--sb-client-privkey\fR string
Private key that the client should use for talking to the OVN database.  Leave empty to use local unix socket.
.TP
\fB\--sb-client-cert\fR string
Client certificate that the client should use for talking to the OVN database.  Leave empty to use local unix socket.
.TP
\fB\--sb-client-cacert\fR string
CA certificate that the client should use for talking to the OVN database.  Leave empty to use local unix socket.
.TP
\fB\--sb-cert-common-name\fR string
The Common Name of the certificate used for TLS server certificate verification.
.TP
\fB\--ovnkube-node-mode\fR string
ovnkube-node operating mode full(default), smart-nic, smart-nic-host (default: "full")
.TP
\fB\--help\fR, \fB\-h\fR
Show help.
.TP
\fB\--version\fR, \fB\-v\fR
Print the version.

.SH "SEE ALSO"
.BR ovn-k8s-overlay (1),
.BR ovn-kube-util (1),
.BR ovn_k8s (5).

.PP
https://github.com/ovn-org/ovn-kubernetes


go-controller/cmd/ovnkube/ovnkube.go
package main

import (
	"context"
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"os/signal"
	"strings"
	"sync"
	"syscall"
	"text/tabwriter"
	"text/template"
	"time"

	"k8s.io/klog/v2"

	goovn "github.com/ebay/go-ovn"
	"github.com/urfave/cli/v2"

	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/metrics"
	ovnnode "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/node"
	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn"
	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/types"
	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"

	kexec "k8s.io/utils/exec"
)

const (
	// CustomAppHelpTemplate helps in grouping options to ovnkube
	CustomAppHelpTemplate = `NAME:
   {{.Name}} - {{.Usage}}

USAGE:
   {{.HelpName}} [global options]

VERSION:
   {{.Version}}{{if .Description}}

DESCRIPTION:
   {{.Description}}{{end}}

COMMANDS:{{range .VisibleCategories}}{{if .Name}}

   {{.Name}}:{{end}}{{range .VisibleCommands}}
     {{join .Names ", "}}{{"\t"}}{{.Usage}}{{end}}{{end}}

GLOBAL OPTIONS:{{range $title, $category := getFlagsByCategory}}
   {{upper $title}}
   {{range $index, $option := $category}}{{if $index}}
   {{end}}{{$option}}{{end}}
   {{end}}`
)

func getFlagsByCategory() map[string][]cli.Flag {
	m := map[string][]cli.Flag{}
	m["Generic Options"] = config.CommonFlags
	m["CNI Options"] = config.CNIFlags
	m["K8s-related Options"] = config.K8sFlags
	m["OVN Northbound DB Options"] = config.OvnNBFlags
	m["OVN Southbound DB Options"] = config.OvnSBFlags
	m["OVN Gateway Options"] = config.OVNGatewayFlags
	m["Master HA Options"] = config.MasterHAFlags
	m["OVN Kube Node flags"] = config.OvnKubeNodeFlags

	return m
}

// borrowed from cli packages' printHelpCustom()
func printOvnKubeHelp(out io.Writer, templ string, data interface{}, customFunc map[string]interface{}) {
	funcMap := template.FuncMap{
		"join":               strings.Join,
		"upper":              strings.ToUpper,
		"getFlagsByCategory": getFlagsByCategory,
	}
	for key, value := range customFunc {
		funcMap[key] = value
	}

	w := tabwriter.NewWriter(out, 1, 8, 2, ' ', 0)
	t := template.Must(template.New("help").Funcs(funcMap).Parse(templ))
	err := t.Execute(w, data)
	if err == nil {
		_ = w.Flush()
	}
}

func main() {
	cli.HelpPrinterCustom = printOvnKubeHelp
	c := cli.NewApp()
	c.Name = "ovnkube"
	c.Usage = "run ovnkube to start master, node, and gateway services"
	c.Version = config.Version
	c.CustomAppHelpTemplate = CustomAppHelpTemplate
	c.Flags = config.GetFlags(nil)

	c.Action = func(c *cli.Context) error {
		return runOvnKube(c)
	}

	ctx := context.Background()

	// trap SIGHUP, SIGINT, SIGTERM, SIGQUIT and
	// cancel the context
	ctx, cancel := context.WithCancel(ctx)
	exitCh := make(chan os.Signal, 1)
	signal.Notify(exitCh,
		syscall.SIGHUP,
		syscall.SIGINT,
		syscall.SIGTERM,
		syscall.SIGQUIT)
	defer func() {
		signal.Stop(exitCh)
		cancel()
	}()
	go func() {
		select {
		case s := <-exitCh:
			klog.Infof("Received signal %s. Shutting down", s)
			cancel()
		case <-ctx.Done():
		}
	}()

	if err := c.RunContext(ctx, os.Args); err != nil {
		klog.Exit(err)
	}
}

func delPidfile(pidfile string) {
	if pidfile != "" {
		if _, err := os.Stat(pidfile); err == nil {
			if err := os.Remove(pidfile); err != nil {
				klog.Errorf("%s delete failed: %v", pidfile, err)
			}
		}
	}
}

func setupPIDFile(pidfile string) error {
	// need to test if already there
	_, err := os.Stat(pidfile)

	// Create if it doesn't exist, else exit with error
	if os.IsNotExist(err) {
		if err := ioutil.WriteFile(pidfile, []byte(fmt.Sprintf("%d", os.Getpid())), 0644); err != nil {
			klog.Errorf("Failed to write pidfile %s (%v). Ignoring..", pidfile, err)
		}
	} else {
		// get the pid and see if it exists
		pid, err := ioutil.ReadFile(pidfile)
		if err != nil {
			return fmt.Errorf("pidfile %s exists but can't be read: %v", pidfile, err)
		}
		_, err1 := os.Stat("/proc/" + string(pid[:]) + "/cmdline")
		if os.IsNotExist(err1) {
			// Left over pid from dead process
			if err := ioutil.WriteFile(pidfile, []byte(fmt.Sprintf("%d", os.Getpid())), 0644); err != nil {
				klog.Errorf("Failed to write pidfile %s (%v). Ignoring..", pidfile, err)
			}
		} else {
			return fmt.Errorf("pidfile %s exists and ovnkube is running", pidfile)
		}
	}

	return nil
}

func runOvnKube(ctx *cli.Context) error {
	pidfile := ctx.String("pidfile")
	if pidfile != "" {
		defer delPidfile(pidfile)
		if err := setupPIDFile(pidfile); err != nil {
			return err
		}
	}

	exec := kexec.New()
	_, err := config.InitConfig(ctx, exec, nil)
	if err != nil {
		return err
	}

	if err = util.SetExec(exec); err != nil {
		return fmt.Errorf("failed to initialize exec helper: %v", err)
	}

	ovnClientset, err := util.NewOVNClientset(&config.Kubernetes)
	if err != nil {
		return err
	}

	master := ctx.String("init-master")
	node := ctx.String("init-node")

	cleanupNode := ctx.String("cleanup-node")
	if cleanupNode != "" {
		if master != "" || node != "" {
			return fmt.Errorf("cannot specify cleanup-node together with 'init-node or 'init-master'")
		}

		if err = ovnnode.CleanupClusterNode(cleanupNode); err != nil {
			return err
		}
		return nil
	}

	if master == "" && node == "" {
		return fmt.Errorf("need to run ovnkube in either master and/or node mode")
	}

	stopChan := make(chan struct{})
	wg := &sync.WaitGroup{}

	var watchFactory factory.Shutdownable
	var masterWatchFactory *factory.WatchFactory
	if master != "" {
		var err error
		// create factory and start the controllers asked for
		masterWatchFactory, err = factory.NewMasterWatchFactory(ovnClientset)
		if err != nil {
			return err
		}
		watchFactory = masterWatchFactory
		var ovnNBClient, ovnSBClient goovn.Client

		if ovnNBClient, err = util.NewOVNNBClient(); err != nil {
			return fmt.Errorf("error when trying to initialize go-ovn NB client: %v", err)
		}

		if ovnSBClient, err = util.NewOVNSBClient(); err != nil {
			return fmt.Errorf("error when trying to initialize go-ovn SB client: %v", err)
		}

		// register prometheus metrics exported by the master
		// this must be done prior to calling controller start
		// since we capture some metrics in Start()
		metrics.RegisterMasterMetrics(ovnNBClient, ovnSBClient)

		ovnController := ovn.NewOvnController(ovnClientset, masterWatchFactory, stopChan, nil, ovnNBClient, ovnSBClient, util.EventRecorder(ovnClientset.KubeClient))
		if err := ovnController.Start(master, wg, ctx.Context); err != nil {
			return err
		}
	}

	if node != "" {
		var nodeWatchFactory factory.NodeWatchFactory
		if masterWatchFactory == nil {
			var err error
			nodeWatchFactory, err = factory.NewNodeWatchFactory(ovnClientset, node)
			if err != nil {
				return err
			}
			watchFactory = nodeWatchFactory
		} else {
			nodeWatchFactory = masterWatchFactory
		}

		if config.Kubernetes.Token == "" {
			return fmt.Errorf("cannot initialize node without service account 'token'. Please provide one with --k8s-token argument")
		}
		// register ovnkube node specific prometheus metrics exported by the node
		metrics.RegisterNodeMetrics()
		start := time.Now()
		n := ovnnode.NewNode(ovnClientset.KubeClient, nodeWatchFactory, node, stopChan, util.EventRecorder(ovnClientset.KubeClient))
		if err := n.Start(wg); err != nil {
			return err
		}
		end := time.Since(start)
		metrics.MetricNodeReadyDuration.Set(end.Seconds())
	}

	// now that ovnkube master/node are running, lets expose the metrics HTTP endpoint if configured
	// start the prometheus server to serve OVN K8s Metrics (default master port: 9409, node port: 9410)
	if config.Kubernetes.MetricsBindAddress != "" {
		metrics.StartMetricsServer(config.Kubernetes.MetricsBindAddress, config.Kubernetes.MetricsEnablePprof)
		if config.OvnKubeNode.Mode != types.NodeModeSmartNICHost && config.Kubernetes.OVNMetricsBindAddress == "" {
			metrics.RegisterOvnMetrics(ovnClientset.KubeClient, master)
		}
	}

	// start the prometheus server to serve OVN Metrics (default port: 9476)
	// Note: for ovnkube node mode smart-nic-host no ovn metrics is required as ovn is not running on the node.
	if config.OvnKubeNode.Mode != types.NodeModeSmartNICHost && config.Kubernetes.OVNMetricsBindAddress != "" {
		metrics.RegisterOvnMetrics(ovnClientset.KubeClient, master)
		metrics.StartOVNMetricsServer(config.Kubernetes.OVNMetricsBindAddress)
	}

	// run until cancelled
	<-ctx.Context.Done()
	close(stopChan)
	watchFactory.Shutdown()
	wg.Wait()
	return nil
}
